### PR TITLE
[CP-4] Add GitHub Actions workflow for build and test

### LINKED ISSUE #4: [CC] Add GitHub Actions workflow for build and test
Define a `Build - Test - Push` workflow having three jobs:

1. `branch`: This job gets the name of the branch being pushed or dispatched to and sets an environment variable based on the branch name (`production` for `master`, `sandbox` for other branches).

2. `test`: This job runs e2e tests for the project using Go and the `go test` command.

3. `build-ami`: This job builds an Amazon Machine Image (AMI) for production (if the push is to the `master` branch) using Packer.

All jobs are configured to run on every push to the repository and also can be triggered manually from the GitHub Actions page using the `workflow_dispatch` event.

## Environments
<ul>
  <li>Sandbox</li>
  <li>Production</li>
</ul>

Both environments have the same following secrets:
<ul>
  <li>AWS_ACCESS_KEY_ID</li>
  <li>AWS_SECRET_ACCESS_KEY</li>
</ul>

## Link to Pull Request

[Link to Pull Request](https://github.com/BeanCloudServices/bcs-cloud-controller/pull/3)

### LINKED ISSUE #5: [CC-5] Add GitHub Actions workflow for build and test
I've created a new branch and added a `.github/workflows` directory with a YAML file that defines the workflow. The workflow includes three jobs:

1. `branch`: This job gets the name of the branch being pushed or dispatched to and sets an environment variable based on the branch name (`production` for `master`, `sandbox` for other branches).

2. `test`: This job runs e2e tests for the project using Go and the `go test` command.

3. `build-ami`: This job builds an Amazon Machine Image (AMI) for production (if the push is to the `master` branch) using Packer.

All jobs are configured to run on every push to the repository and also can be triggered manually from the GitHub Actions page using the `workflow_dispatch` event.

## Environments
<ul>
  <li>Sandbox</li>
  <li>Production</li>
</ul>

Both environments have the same following secrets:
<ul>
  <li>AWS_ACCESS_KEY_ID</li>
  <li>AWS_SECRET_ACCESS_KEY</li>
</ul>

## Link to Pull Request

[Link to Pull Request](https://github.com/BeanCloudServices/bcs-cloud-controller/pull/3)

### PR DESCRIPTION
**Pull request by henrypham67**
This closes https://github.com/BeanProjects/cloud-platform/issues/4

### CODE REVIEW

**Comment by henrypham67** [comment]
GitHub Actions knows which branch to fetch without having input

image

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
+name: Packer Build
```
Comment (binh-le-thanh-mox) [inline .github/workflows/ci.yml:1]:
```suggestion
name: Build - Test - Push

```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+          else
+             echo "env_name=sandbox" >> $GITHUB_OUTPUT
+          fi
+    outputs:
+      env_name: ${{ steps.branch-check.outputs.env_name }}
+
+  test:
+    runs-on: ubuntu-latest
+    needs: [branch]
+    environment: ${{ needs.branch.outputs.env_name }}
```
Comment (binh-le-thanh-mox) [inline .github/workflows/ci.yml:33]: Pushing test AMI to production env does not make any sense

**Review by binh-le-thanh-mox — CHANGES_REQUESTED** [review]
Please rename `.github/workflows/ci.yml`
to `.github/workflows/build-test-push.yml`

File: .github/workflows/build-test-push.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+      - name: Set up Go
+        uses: actions/setup-go@v4
+        with:
+          go-version: '1.20'
+
+      - name: Test
+        run: go test ./test/e2e/*
+
+  build:
```
Comment (peterbean410) [inline .github/workflows/build-test-push.yml]:
```suggestion
  build-ami:

```

File: .github/workflows/build-test-push.yml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      - name: Set up Go
+        uses: actions/setup-go@v4
+        with:
+          go-version: '1.20'
+
+      - name: Test
+        run: go test ./test/e2e/*
+
+  build-ami:
+
```
Comment (peterbean410) [inline .github/workflows/build-test-push.yml:50]:
```suggestion
```

File: .github/workflows/build-test-push.yml [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+          echo "Running on branch ${{ github.ref }}"
+          if [ "${{ github.ref }}" == "refs/heads/master" ]; then
+            echo "env_name=production" >> $GITHUB_OUTPUT
+          else
+             echo "env_name=sandbox" >> $GITHUB_OUTPUT
+          fi
+    outputs:
+      env_name: ${{ steps.branch-check.outputs.env_name }}
+
+  test:
```
Comment (peterbean410) [inline .github/workflows/build-test-push.yml:31]:
```suggestion
  test:
    name: Run Tests
```

File: .github/workflows/build-test-push.yml [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    name: Run Tests
```
Comment (peterbean410) [inline .github/workflows/build-test-push.yml]:
```suggestion
    name: Run tests

```

File: .github/workflows/build-test-push.yml [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+          if [ "${{ github.ref }}" == "refs/heads/master" ]; then
+            echo "env_name=production" >> $GITHUB_OUTPUT
+          else
+             echo "env_name=sandbox" >> $GITHUB_OUTPUT
+          fi
+    outputs:
+      env_name: ${{ steps.branch-check.outputs.env_name }}
+
+  test:
+    name: Run tests
```
Comment (peterbean410) [inline .github/workflows/build-test-push.yml]:
```suggestion
    name: Test

```

File: .github/workflows/build-test-push.yml [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      - name: Set up Go
+        uses: actions/setup-go@v4
+        with:
+          go-version: '1.20'
+
+      - name: Test
+        run: go test ./test/e2e/*
+
+  build-ami:
+    name: Packer Build
```
Comment (peterbean410) [inline .github/workflows/build-test-push.yml]:
```suggestion
    name: Build AMI

```

**Pull request merged**